bluesky_browser/suitcase.py

```diff
@@ -0,0 +1,389 @@
+        self.setGeometry(*self.geometry)
+        self.main_layout = QFormLayout()
+
+        # add an 'enable exporter' checkbox.
+        self.enable_checkbox = QCheckBox('enable exporter', self)
+        self.main_layout.addRow('', self.enable_checkbox)
+
+        # set the trait widgets for `file_prefix` and `directory`.
+        self._create_trait_widget('directory', QLineEdit, 'setText',
+                                  'textChanged')
+        self._create_trait_widget('file_prefix', QLineEdit, 'setText',
+                                  'textChanged')
+
+    def show_parameters(self, parameters=[]):
+        '''Walks through ``self.parameters`` then ``parameters`` showing the
+           widgets.'''
+
+        # add any parameters from self.parameters, if they are unique.
+        combined_parameters = []
+        for parameter in self.parameters:
+            if parameter not in combined_parameters:
+                combined_parameters.append(parameter)
+        # add any parameters from the kwarg parameters, if they are unique.
+        for parameter in parameters:
+            if parameter not in combined_parameters:
+                combined_parameters.append(parameter)
+
+        # step through each parameter and add the widget to the layout.
+        for parameter in combined_parameters:
+            # collect the trait and widget_type objects.
+            trait_widget = getattr(self, f'{parameter}_widget')
+            self.main_layout.addRow(parameter, trait_widget)
+
+        # set ``self.main_layout`` as the layout
+        self.setLayout(self.main_layout)
+
+    def _create_trait_widget(self, trait_name, widget_type, slot_name,
+                             signal_name):
+        '''Creates an input widget for a trait based on the input parameters.
+
+        This creates an input widget for the trait specified by ``trait_name``
+        using ``widget_type`` as the type. It also ensures that the current
+        value of ``trait_name`` is set to the widgets ``slot_name`` and that
+        ``trait_name`` is connected to the widgets ``signal_name``.
+
+        Parameters
+        ----------
+        trait_name : str
+            The name of the trait as a string, i.e. the trait is found at
+            ``self.trait_name``.
+        widget_type : obj
+            The pyqt5 widget class that is to be used to create the input
+            widget, e.g. ``PyQt5.QtWidgets.QLineEdit``
+        slot_name : str
+            The name of the ``PyQt5.QtCore.pyqtSlot from ``widget_type`` to set
+            the current value too, e.g. ``PyQt5.QtWidget.QlineEdit.setText``.
+        signal_name : str
+            The name of the ``PyQt5.QtCore.pyqtSignal from ``widget_type`` to
+            connect the trait too, e.g. ``PyQt5.QtWidget.QlineEdit.setText``.
+        '''
+
+        # reference the trait and create the trait widget
+        trait = getattr(self, trait_name)
+        setattr(self, f'{trait_name}_widget', widget_type())
```

Comment:
How about defining `self._widgets = {}` in `__init__` and then doing `self._widgets[trait_name] = widget_type()`? That would avoid monkey-patching and string-munging.